[mlir][bufferization] Support custom types at function boundaries

andrey-golubev opened the pull request:
Support custom types (3/N): allow custom tensor and buffer types in function signatures and at call-sites. This is one of the major building blocks to move in the direction of module-level one-shot-bufferization support.

To achieve this, `BufferizationOptions::FunctionArgTypeConverterFn` callback is converted to work with tensor-like and buffer-like types, instead of the builtin counterparts. The default behavior for builtins remains unchanged, while custom types by default go through `TensorLikeType::getBufferType()` which is a general conversion interface.

andrey-golubev committed:
[mlir][bufferization] Support custom types at function boundaries

Support custom types (3/N): allow custom tensor and buffer types in
function signatures and at call-sites. This is one of the major building
blocks to move in the direction of module-level one-shot-bufferization
support.

In order to enable this, TensorLikeType is extended with a new interface
method that is invoked solely within the function boundary
bufferization.

andrey-golubev commented on mlir/lib/Dialect/Bufferization/IR/BufferizationDialect.cpp:
```diff
@@ -87,6 +87,19 @@ struct BuiltinTensorExternalModel
+
+    return cast<BufferLikeType>(options.functionArgTypeConverterFn(
+        tensorType, *memSpace, funcOp, options));
+  }
```
side-note: there is a new behavior related to mem-space checking. Before, the logic would unconditionally access the value returned by `defaultMemorySpaceFn()`, assuming "no error".

I decided to make the getBufferTypeAtFunctionBoundary() signature aligned with normal getBuffer(). And thus checking for memspace existence seems like a reasonable addition.

Technically, we don't have to return a failure here, but instead can just ignore this case completely (as the original logic goes). Moreover, all current users of this API expect "always success".

andrey-golubev commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferizationTypeInterfaces.h:
```diff
@@ -13,6 +13,7 @@
 // Bufferization Type Interfaces
 //===----------------------------------------------------------------------===//
 
+#include "mlir/Dialect/Func/IR/FuncOps.h" // to access mlir::func::FuncOp
```
i didn't think of this, yes, i think we should.

andrey-golubev commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferizationTypeInterfaces.td:
```diff
@@ -43,6 +43,18 @@ def Bufferization_TensorLikeTypeInterface
         "::mlir::bufferization::BufferLikeType":$bufferType,
         "::llvm::function_ref<::mlir::InFlightDiagnostic()>":$emitError)
+    >,
+    InterfaceMethod<[{
```
main issue with this new interface method is that I can't seem to provide a default implementation. i'd go with `new method` == `getBufferType` by default but it fails to compile since builtin types get this interface dynamically attached.

matthias-springer replied:
> i'm actually intrigued why the FunctionArgTypeConverterFn needs it

I don't know either. But it could be useful to get more context to do the type conversion.

> thought of the interface here because this seems to be aligned with what we've discussed before:

The problem that I'm seeing is that there are now two ways to configure the type conversion, which is a bit confusing. If we make this interface-based, we should delete the options callback. But I think that's not possible because you wouldn't be able to configure identity/fully dynamic layout maps at the function boundary.

Function boundaries are always a bit special, so I think having an options callback is desirable also for custom tensor/buffer types.

> doing this via options practically means users have to create their own "pass" (i.e. bufferizeOp()'s options have to be seeded with different functions that aren't really configurable outside of C++)

That is the case in every production compiler that I'm aware of. This is not the only flag that cannot be customized as a pass flag. `AllocationFn`, `MemCpyFn` is another one.

andrey-golubev replied:
> But I think that's not possible because you wouldn't be able to configure identity/fully dynamic layout maps at the function boundary.

I'm not sure I understand this part, can you explain more? Do you mean the `void setFunctionBoundaryTypeConversion(LayoutMapOption layoutMapOption);` setter that makes the conversion `LayoutMapOption` value-dependent? I think, yes, if this is a hard blocker, an option callback must stay - passing `LayoutMapOption` via interfaces is probably a pain (if possible without hacks or a significant redesign).

> Function boundaries are always a bit special, so I think having an options callback is desirable also for custom tensor/buffer types.

Fair point. Let's go with this then. I think I could do it a bit differently: for `TensorType` (current API), we keep the logic "as is", otherwise - for `TensorLikeType` - we always call it's `getBufferType()`? This would preserve single conversion "interface" - the options callback and also give us some sane default behavior for any type (i.e. i wouldn't need to create --test-one-shot-bufferize pass yet).

> That is the case in every production compiler that I'm aware of. This is not the only flag that cannot be customized as a pass flag. AllocationFn, MemCpyFn is another one.

Right, also in our case that's the way it is (fwiw, this is probably inevitable). I was a bit lazy to have a test pass in MLIR for custom types though (i'll probably have one anyway but later: to test how builtins with `tensor encoding` -> `memref layout` behave, this is one of the important things we use everywhere in our downstream).

andrey-golubev replied:
removed this interface method.

matthias-springer replied:
> I'm not sure I understand this part, can you explain more?

Interface method implementation cannot be overridden by the user. In the interface implementation you have to make a choice whether to convert a tensor type to a memref with identity layout or fully dynamic layout map. We want users to be able to customize behavior, therefore, there must be some kind of hook in the `BufferizationOptions`.

andrey-golubev committed:
Rely on BufferizationOptions::FunctionArgTypeConverterFn

Transform FunctionArgTypeConverterFn into a tensor-like -> buffer-like
converter so that it could be used as a generic function boundary
conversion utility.

andrey-golubev commented on mlir/lib/Dialect/Bufferization/Transforms/FuncBufferizableOpInterfaceImpl.cpp:
```diff
@@ -49,29 +49,47 @@ void FuncAnalysisState::startFunctionAnalysis(FuncOp funcOp) {
+    return *options.defaultMemorySpaceFn(tensorType);
+  }
+  return nullptr;
+}
```
note: i guess we could also make `defaultMemorySpaceFn` work with tensor-like type. but i'd prefer to do it separately to simplify the updates for the users.

otoh, TensorLikeType::getBufferType() already has access to the tensor type and can infer memory space from it. I am not sure why we need it explicitly at all.

andrey-golubev commented on mlir/include/mlir/Dialect/Bufferization/IR/BufferizableOpInterface.h:
```diff
@@ -335,10 +335,12 @@ struct BufferizationOptions {
+  /// Type converter from tensors to buffers. This type converter is used to
+  /// determine bufferized function argument and result types.
+  ///
+  /// By default, if tensor is a (builtin) tensor type, a type converter that
```
ack. i think this is original wording (just slightly refurbished).

andrey-golubev replied:
changed to "type conversion".  also for `UnknownTypeConverterFn`.

andrey-golubev committed:
Change documentation wording for type converters
